internal/tfdiags/diagnostics.go

```diff
@@ -101,6 +102,23 @@ func (diags Diagnostics) Append(new ...interface{}) Diagnostics {
 	return diags
 }
 
+func (diags Diagnostics) Merge(other Diagnostics) Diagnostics {
+	if len(diags) == 0 {
+		return other
+	}
+
+	for _, d := range diags {
+		for _, o := range other {
+			if reflect.DeepEqual(d, o) {
```

Comment:
I was thinking about adding Equal method but `Diagnostic` interface has many structs/interfaces that also has other struct/interface dependencies. For the comparison check it would become a overkill for adding that much `Equal()` functions to many structs. That's why I have used `reflect.DeepEqual` here

Reply:
`reflect.DeepEqual` feels a bit of an overkill here. Also, it could fail with wrapped diags.

Would you mind adapting the logic in [ConsolidateWarnings](https://github.com/opentofu/opentofu/blob/main/internal/tfdiags/consolidate_warnings.go#L25) here?

I think we can treat two diags as duplicates if:
* their `Severity`s are equal
* their `Source.Subject`s are present and the whole `Source`s are equal
* their `Description`s are equal

I would also check if `DoNotConsolidateDiagnostic` function allows merging the diag.

Reply:
I've gone back and forth on this.  We know we are using overlapping code paths to produce duplicate diagnostics.  I'm ok with this being a bit stricter than ConsolidateWarnings.  I think the DeepEqual is probably a bit overkill, but @ollevche's list makes sense, perhaps with the Ranges added as well.

Reply:
I also think that DoNotConsolidateDiagnostic can be ignored for this particular function, as it is a special purpose function only for this scenario.

Reply:
Yep, I agree `DoNotConsolidateDiagnostic` check is optional here.

> perhaps with the Ranges added as well.

`Source`s should be equal and `Source.Subject`s (ranges) should be non-zero.

Reply:
Added Equal() methods to used structs and refactored the function